### PR TITLE
8323900: Avoid calling os::init_random() in CDS static dump

### PR DESCRIPTION
**Pull request by iklam**
The purpose of the PR is to avoid modifying the global JVM state while dumping the CDS archive.

When updating the identity hashcode for archived Symbols, call `ArchiveBuilder::current()->entropy()` instead of `os::random()`. As a result, CDS no longer needs to call `os::init_random()` with a deterministic seed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8323900](https://bugs.openjdk.org/browse/JDK-8323900): Avoid calling os::init_random() in CDS static dump (**Enhancement** - P4)

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18728/head:pull/18728` \
`$ git checkout pull/18728`

Update a local copy of the PR: \
`$ git checkout pull/18728` \
`$ git pull https://git.openjdk.org/jdk.git pull/18728/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18728`

View PR using the GUI difftool: \
`$ git pr show -t 18728`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18728.diff">https://git.openjdk.org/jdk/pull/18728.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18728#issuecomment-2048011645)

### CODE REVIEW

**Review by tstuefe — APPROVED** [review]
Looks good!

File: src/hotspot/share/cds/archiveBuilder.hpp [hunk 1]
```diff
@@ -212,6 +212,7 @@ class ArchiveBuilder : public StackObj {
   ResizeableResourceHashtable<address, address, AnyObj::C_HEAP, mtClassShared> _buffered_to_src_table;
   GrowableArray<Klass*>* _klasses;
   GrowableArray<Symbol*>* _symbols;
+  unsigned int _entropy_seed;
```
Comment (ashu-mehra) [inline src/hotspot/share/cds/archiveBuilder.hpp:215]: Shouldn't it be a static const?

**Comment by tstuefe** [comment]
Thinking about this, since global entropy (archived object ihashes) sneak into archives whether we use local seeds or not, maybe we should not bother with such a patch. 

In other words, if global state affects the archive anyway, we may just as well roll with it.

See https://github.com/openjdk/jdk/pull/18735

File: src/hotspot/share/cds/archiveBuilder.hpp (hunk 1 above)
Comment (tstuefe) [reply src/hotspot/share/cds/archiveBuilder.hpp:215]: I think it is meant to be a member of the (one and only existing) archive builder.

**Comment by iklam** [comment]
> Thinking about this, since global entropy (archived object ihashes) sneak into archives whether we use local seeds or not, maybe we should not bother with such a patch.
> 
> In other words, if global state affects the archive anyway, we may just as well roll with it.
> 
> See #18735

In CDS, we intend to be as much independent of the global JVM state as possible. For example, since [JDK-8296344](https://bugs.openjdk.org/browse/JDK-8296344), we no longer make a copy of the the archived heap objects in the actual Java heap.

The intention of this PR is the same -- the contents of archived Symbols should not depend on the value of the os::random() seed.

In #18735 you found that some other contents of the CDS archive depend on the JVM's os::random() seed. That may be something we want to fix separately. In any case, that's not a reason to not proceed with this PR.

**Comment by tstuefe** [comment]
> > Thinking about this, since global entropy (archived object ihashes) sneak into archives whether we use local seeds or not, maybe we should not bother with such a patch.
> > In other words, if global state affects the archive anyway, we may just as well roll with it.
> > See #18735
> 
> In CDS, we intend to be as much independent of the global JVM state as possible. For example, since [JDK-8296344](https://bugs.openjdk.org/browse/JDK-8296344), we no longer make a copy of the the archived heap objects in the actual Java heap.
> 
> The intention of this PR is the same -- the contents of archived Symbols should not depend on the value of the os::random() seed.
> 
> In #18735 you found that some other contents of the CDS archive depend on the JVM's os::random() seed. That may be something we want to fix separately. In any case, that's not a reason to not proceed with this PR.

Okay. Thinking about this, an isolated seed is always better, since it provides safety against concurrent uses of os::random (which can happen even at initialization time). 

So my approval stands.

**Review by calvinccheung — APPROVED** [review]
Looks good.

**Comment by iklam** [comment]
Thanks @tstuefe and @calvinccheung for the review.
/integrate

**Pull request closed without merging**